src/librustc/ty/util.rs

```diff
@@ -676,37 +676,43 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
 
     /// Check whether a type is representable. This means it cannot contain unboxed
     /// structural recursion. This check is needed for structs and enums.
-    pub fn is_representable(&'tcx self, tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span)
+    pub fn is_representable(&'tcx self,
+                            tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                            sp: Span)
                             -> Representability {
 
         // Iterate until something non-representable is found
         fn find_nonrepresentable<'a, 'tcx, It>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                                               sp: Span,
                                                seen: &mut Vec<Ty<'tcx>>,
                                                iter: It)
                                                -> Representability
-        where It: Iterator<Item=Ty<'tcx>> {
+        where It: Iterator<Item=(Ty<'tcx>, Span)> {
             iter.fold(Representability::Representable,
-                      |r, ty| cmp::max(r, is_type_structurally_recursive(tcx, sp, seen, ty)))
+                      |r, (ty, sp)| cmp::max(r, is_type_structurally_recursive(tcx, sp, seen, ty)))
         }
 
         fn are_inner_types_recursive<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span,
                                                seen: &mut Vec<Ty<'tcx>>, ty: Ty<'tcx>)
                                                -> Representability {
             match ty.sty {
                 TyTuple(ref ts, _) => {
-                    find_nonrepresentable(tcx, sp, seen, ts.iter().cloned())
+                    find_nonrepresentable(tcx, seen, ts.iter().cloned().map(|t| (t, sp)))
                 }
                 // Fixed-length vectors.
                 // FIXME(#11924) Behavior undecided for zero-length vectors.
                 TyArray(ty, _) => {
                     is_type_structurally_recursive(tcx, sp, seen, ty)
                 }
                 TyAdt(def, substs) => {
-                    find_nonrepresentable(tcx,
-                                          sp,
+                    let spans = def.all_fields().map(|f| {
+                        tcx.hir.span_if_local(f.did).unwrap_or(DUMMY_SP)
+                    }).collect::<Vec<Span>>();
+                    match find_nonrepresentable(tcx,
                                           seen,
-                                          def.all_fields().map(|f| f.ty(tcx, substs)))
+                                          def.all_fields().map(|f| (f.ty(tcx, substs), tcx.hir.span_if_local(f.did).unwrap_or(DUMMY_SP)))) {
+                        Representability::SelfRecursive(_) => Representability::SelfRecursive(spans),
```

Comment:
Yeah, the last commit is mostly me trying to modify the logic here as little as possible while adding spans. It seems to me that I'll have to take some time to rewrite this to properly keep a stack.